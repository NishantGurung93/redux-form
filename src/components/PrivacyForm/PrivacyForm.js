import { useDispatch } from "react-redux";
import { useHistory } from "react-router-dom";
import { Form, Formik } from "formik";
import Checkbox from "../Checkbox";
import { submitPrivacy } from "../../slices/privacySlice";

const PrivacyForm = () => {
  const dispatch = useDispatch();
  const history = useHistory();

  const initialValues = {
    trayEmail: '',
    otherEmail: '',
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={(values) => {
        dispatch(submitPrivacy(values));
        history.push('/confirmation');
      }}
    >
      <Form>
        <Checkbox
          label="Receive updates about Tray.io product by email"
          name="trayEmail"
          type="checkbox"
        />

        <Checkbox
          label="Receive communication by email for other products created by the Tray.io team"
          name="otherEmail"
          type="checkbox"
        />
        <button type="submit" data-testid="submit-button">Submit</button>
      </Form>
    </Formik>
  )
};

export default PrivacyForm;
