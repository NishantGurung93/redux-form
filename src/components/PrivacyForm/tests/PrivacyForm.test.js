import { fireEvent, render } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import PrivacyForm from "..";

const dispatch = jest.fn();
const historyMock = { push: jest.fn(), location: {}, listen: jest.fn() };

const mockStore = {
  value: {},
  useDispatch: jest.fn(),
  dispatch,
  getState: jest.fn(),
  subscribe: jest.fn(),
}

const renderComponent = () => render(
  <Provider store={mockStore}>
    <Router history={historyMock}>
      <PrivacyForm />
    </Router>
  </Provider>
);

describe('Header component', () => {
  it('should render the component', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  describe('valid form behaviour', () => {
    beforeEach(async () => {
      const { getByTestId } = renderComponent();

      const trayEmail = getByTestId('trayEmail-checkbox');
      fireEvent.click(trayEmail)

      const otherEmail = getByTestId('otherEmail-checkbox');
      fireEvent.click(otherEmail)

      const submitButton = getByTestId('submit-button');

      await act(async () => {
        fireEvent.click(submitButton);
      });
    });

    it('should call submitDetails', () => {
      expect(dispatch).toHaveBeenCalledWith({
        payload: {
          trayEmail: true,
          otherEmail: true,
        },
        type: 'privacy/submitPrivacy',
      });
    });

    it('should call useHistory hook with "/confirmation"', () => {
      expect(historyMock.push).toHaveBeenCalledWith('/confirmation');
    });
  });
});
