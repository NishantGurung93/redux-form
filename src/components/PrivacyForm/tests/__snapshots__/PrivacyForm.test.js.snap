// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render the component 1`] = `
<div>
  <form
    action="#"
  >
    <div
      class="checkbox-wrapper"
    >
      <input
        data-testid="trayEmail-checkbox"
        id="trayEmail"
        name="trayEmail"
        type="checkbox"
        value=""
      />
      <label
        for="trayEmail"
      >
        Receive updates about Tray.io product by email
      </label>
    </div>
    <div
      class="checkbox-wrapper"
    >
      <input
        data-testid="otherEmail-checkbox"
        id="otherEmail"
        name="otherEmail"
        type="checkbox"
        value=""
      />
      <label
        for="otherEmail"
      >
        Receive communication by email for other products created by the Tray.io team
      </label>
    </div>
    <button
      class="submit-button"
      data-testid="submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
